chalk-engine/src/logic.rs

```diff
@@ -238,23 +236,35 @@ impl<C: Context> Forest<C> {
     /// In terms of the NFTD paper, this corresponds to the *Program
     /// Clause Resolution* step being applied eagerly, as many times
     /// as possible.
-    fn push_initial_strands(&mut self, context: &impl ContextOps<C>, table: TableIndex) {
+    fn build_table(
+        context: &impl ContextOps<C>,
+        table_idx: TableIndex,
+        goal: C::UCanonicalGoalInEnvironment,
+    ) -> Table<C> {
         // Instantiate the table goal with fresh inference variables.
-        let table_goal = self.tables[table].table_goal.clone();
+        let table = Table::<C>::new(goal.clone(), context.is_coinductive(&goal));
+        let table_goal = table.table_goal.clone();
         let (infer, subst, environment, goal) = context.instantiate_ucanonical_goal(&table_goal);
-        self.push_initial_strands_instantiated(context, table, infer, subst, environment, goal);
+        Self::push_initial_strands_instantiated(
+            context,
+            table_idx,
+            table,
+            infer,
+            subst,
+            environment,
+            goal,
+        )
     }
 
     fn push_initial_strands_instantiated(
-        &mut self,
         context: &impl ContextOps<C>,
-        table: TableIndex,
+        table_idx: TableIndex,
+        mut table: Table<C>,
```

Comment:
I am not sure, if it wont be performance regression to take initialized table and then return it in result instead of in place mutate it (I find it is not trivial to implement such an optimization, and I don't know how rustc handles this), so I am debating on leaving it as `table: &mut Table<C>`, on the other hand I like it this way better... So I will just do what I would be told to do in this case, but here is my proposal.